cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/writer/BasicLineageWriterTest.java

```diff
@@ -0,0 +1,116 @@
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.data.runtime.DataSetsModules;
+import co.cask.cdap.data2.dataset2.DatasetDefinitionRegistryFactory;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DatasetFrameworkTestUtil;
+import co.cask.cdap.data2.dataset2.DefaultDatasetDefinitionRegistry;
+import co.cask.cdap.data2.dataset2.InMemoryDatasetFramework;
+import co.cask.cdap.data2.metadata.lineage.AccessType;
+import co.cask.cdap.data2.metadata.lineage.LineageStore;
+import co.cask.cdap.data2.metadata.publisher.MetadataChangePublisher;
+import co.cask.cdap.data2.metadata.publisher.NoOpMetadataChangePublisher;
+import co.cask.cdap.data2.metadata.service.BusinessMetadataStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.metadata.MetadataRecord;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.AbstractModule;
+import com.google.inject.Injector;
+import com.google.inject.assistedinject.FactoryModuleBuilder;
+import com.google.inject.name.Names;
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+/**
+ * Tests BasicLineageWriter
+ */
+public class BasicLineageWriterTest {
+  @ClassRule
+  public static DatasetFrameworkTestUtil dsFrameworkUtil = new DatasetFrameworkTestUtil();
+
+  @Test
+  public void testWrites() throws Exception {
+    Injector injector = getInjector();
+    BusinessMetadataStore businessMetadataStore = injector.getInstance(BusinessMetadataStore.class);
+    LineageStore lineageStore = injector.getInstance(LineageStore.class);
+    LineageWriter lineageWriter = new BasicLineageWriter(businessMetadataStore, lineageStore);
+
+    // Define entities
+    Id.Program program = Id.Program.from("default", "app", ProgramType.FLOW, "flow");
+    Id.Stream stream = Id.Stream.from("default", "stream");
+    Id.Run run1 = new Id.Run(program, RunIds.generate(10000).getId());
+    Id.Run run2 = new Id.Run(program, RunIds.generate(20000).getId());
+
+    // Tag stream
+    businessMetadataStore.addTags(stream, "stag1", "stag2");
+    // Write access for run1
+    lineageWriter.addAccess(run1, stream, AccessType.READ);
+    // Assert tags on stream
+    Assert.assertEquals(
+      ImmutableSet.of(
+        new MetadataRecord(stream, ImmutableMap.<String, String>of(), ImmutableSet.of("stag1", "stag2")),
+        new MetadataRecord(program.getApplication(), ImmutableMap.<String, String>of(), ImmutableSet.<String>of()),
+        new MetadataRecord(program, ImmutableMap.<String, String>of(), ImmutableSet.<String>of())),
+      lineageStore.getRunMetadata(run1));
+
+    // Add another tag to stream
+    businessMetadataStore.addTags(stream, "stag3");
+    // Write access for run1 again
+    lineageWriter.addAccess(run1, stream, AccessType.READ);
+    // The write should be no-op, and metadata for run1 should not be updated
```

Comment:
I think we want to record the tags the first time a Dataset is accessed, because that's when the Dataset code is loaded.